mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp

```diff
@@ -145,22 +145,33 @@ Type getResultType(Operation *convOp, Value outArg) {
   return nullptr;
 }
 
+static int64_t getUtilityVectorizationLen(ShapedType shape) {
+  int64_t numElems = shape.getNumElements();
+  constexpr int64_t kMaxVectorOpLen = 4; // words
+  int64_t elemsPerWord = (32 / shape.getElementTypeBitWidth());
+  return math_util::gcd(numElems, kMaxVectorOpLen * elemsPerWord);
+}
+
 /// Create an elementwise utility kernel.
 /// The callback has type (builder, location, collapsedBuffers, coordinate).
 /// Note: you are expected to handle out of bounds, such as by using
 /// rock.buffer_store
+template <typename OpType>
 LogicalResult createElementwiseLoop(
-    OpBuilder &b, Location loc, Operation *convOp, ValueRange memrefs,
+    OpBuilder &b, Location loc, OpType kernelOp, ValueRange memrefs,
     int64_t vectorLen,
     function_ref<void(OpBuilder &, Location, ValueRange, Value)> emitBodyFunc) {
-  uint32_t blockSize =
-      convOp->getAttrOfType<IntegerAttr>("utilityBlockSize").getInt();
-  // Now, this convOp is confirmed as a utility kernel, deleting params.
-  convOp->removeAttr("params");
-  int64_t elemsPerThread =
-      convOp->getAttrOfType<IntegerAttr>("elems_per_thread").getInt();
+  if (!kernelOp.getBlockSize().has_value())
+    return kernelOp.emitOpError("block size not defined for utility kernel");
+  if (!kernelOp.getGridSize().has_value())
+    return kernelOp.emitOpError("grid size not defined for utility kernel");
+  if (!kernelOp.getElemsPerThread().has_value())
+    return kernelOp.emitOpError(
+        "elemsPerThread not defined fer utility kernel");
+  uint32_t blockSize = *kernelOp.getBlockSize();
+  int64_t elemsPerThread = kernelOp.getElemsPerThread()->getSExtValue();
   if (elemsPerThread % vectorLen != 0)
-    return convOp->emitOpError("Unevenly vectorized elementwise kernel");
+    return kernelOp.emitOpError("unevenly vectorized elementwise kernel");
```

Comment:
I think this can be better handled when we populate the vectorization length. Make it into a for loop such that as long as vectorization length is not divisible, cut it to half until it is possible.

Reply:
Well, the vectorization length getter will already ensure this invariant - here is another layer of sanity-checking

Reply:
If I read it right `getUtilityVectorizationLen()` will make its decision mostly on the type width, but didn't take `elemsPerThread` into account. On the other hand, `elemsPerThread` is a very flexible value because it get passed in directly as a op argument. If the externally passed in `elemsPerThread` becomes less than 4 word then this can be invalid.

Reply:
... yep, I should stick `elemsPerThread` into this function. You're right.